### DISCUSSION TITLE
DNS challenge Caddy su Qnap

### DISCUSSION THREAD

**Discussion by skybarberom** [indomus/forum#1886] [HUB: Home Assistant]
Ciao, 
sto seguendo la guida per esporre homeassistant core tramite caddy su container/QNAP (vorrei evitare di pagare nabucasa ora che è aumentato di molto).
Mi è chiaro quanto c'è nella giuda e, facendo un paio di prove, aprendo la 443 sul router ricevo i certificati di prova.
Vorrei però evitare di aprire la 443 sul router e quindi fare dns challenge utilizzando (per semplicità invece di un mio dominio visto che lo trattate nella guida) duckdns.
Come integrate il plugin duckdns in caddy conteinerizzato? Trovo i repo del codice go, ma come lo inserisco nel container? Devo abbandonare l'interfaccia grafica della container station e fare a mano la build con l'aggiunta? Sarebbe bello avere un aiuto e/o in futuro una guida.
Aprire la 443 o la 80 non è che mi piaccia molto!
Grazie !

**Comment by snakuzzo** [comment]
Ciao.
Devi semplicemente fare la build con un tuo Dockerfile specificando i moduli aggiuntivi che ti occorrono. Non so se esiste già un'immagine già pronta, ma se può esserti utile io ne ho fatta una "pronta" con i moduli per le DNS challenge di Cloudflare e DuckDNS

Puoi trarre spunto dal progetto, guardando appunto il Dockerfile

https://github.com/snakuzzo/caddy-cloudflare-duckdns

Oppure puoi scaricare l'immagine da qui...

ghcr.io/snakuzzo/caddy-cloudflare-duckdns:latest

Per il Caddyfile puoi trarre spunto da qui...

```
https://tuodoominio.duckdns.org:9000 {
        import https_header
        reverse_proxy 192.168.1.10:8123
    # CLOUDFLARE
        #tls tuamail@gmail.com {
        #       dns cloudflare tuotoken
        #}
    # DUCKDNS
        tls tuamail@gmail.com {
                dns duckdns tuotoken
        }
```

**Comment by skybarberom** [comment]
Ciao,

grazie per il dockerfile. L'ho scaricato nel QNAP e ho lanciato la build e me la ritrovo correttamente tra le immagini disponibili nell'interfaccia grafica.

Ho integrato il caddyfile come dicevi

```

(https_header) {
  header {
    Strict-Transport-Security "max-age=31536000; includeSubdomains"
    X-XSS-Protection "1; mode=block"
    X-Content-Type-Options "nosniff"
    X-Frame-Options "SAMEORIGIN"
    Referrer-Policy "same-origin"
  }
}

https://MIONOME.duckdns.org:9123 { 
  import https_header
  reverse_proxy http://192.168.0.191:8123
    # DUCKDNS
    tls MIAMAIL@gmail.com {
    dns duckdns xxxxx-xxxxx-xxxx-xxxx-xxxxxxxx
    }
}

```

ma mi ritrovo questo errore sulla parte tls

```

2022/09/04 20:50:49.933 INFO    using provided configuration    {"config_file": "/etc/caddy/Caddyfile", "config_adapter": "caddyfile"}
run: adapting config using caddyfile: parsing caddyfile tokens for 'tls': /etc/caddy/Caddyfile:16 - Error during parsing: getting module named 'dns.providers.duckdns': module not registered: dns.providers.duckdns

```

come se ci fosse un errore di formattazione da qualche parte (ma c'è anche quel module not registered che mi fa pensare)

Hai qualche suggerimento?

Grazie!

**Reply by snakuzzo** [reply]
non è un errore di formattazione... dice...

```
module not registered: dns.providers.duckdns
```

Controlla la build

**Reply by skybarberom** [reply]
Ciao,

non sono espertissimo sui docker, io ho semplicemente lanciato questi comandi per scaricare il docker file e fare la build:

```

cd /tmp && curl -LO https://github.com/snakuzzo/caddy-cloudflare-duckdns/archive/refs/heads/latest.zip

unzip latest.zip

cd /tmp/caddy-cloudflare-duckdns-latest/     

docker build .

```

La build sembra essere stata creata con successo e mi ritrovo nell'interfaccia grafica due immagini:

```
caddy
builder
0ce17ec8cd80
328.57 MB
2022/08/10 02:26:07

caddy
latest
743ed904eb39
42.87 MB
2022/08/10 02:26:01

```

Io ho lanciato la latest (è corretto?) con il caddyfile che ho scritto sopra e con il solito comando di lancio che utilizzavo:

```
caddy run --config /etc/caddy/Caddyfile --adapter Caddyfile
```

Mi sto perdendo qualcosa?

**Reply by snakuzzo** [reply]
Scusa eh...ma che diavolo c'entra l'ultimo comando ? stai lanciando caddy come servizio e non stai creando il container docker. Stai facendo parecchia confusione.
Se hai creato correttamente l'immagine, assicurati prima di aver fermato e rimosso il vecchio container di caddy (per evitare conflitti con le porte utilizzate) e poi devi creare il container nuovo (magari dandogli un nome diverso).
Ricordati di montare anche una directory data dove memorizzare i certificati, se non lo fai ogni volta ricrei il container farai richiesta di nuovi certificati e questo potrebbe comportarti qualche problema, visto che le richieste che puoi fare sono "limitate" nel tempo.

Una roba simile...

Crei la struttura di directory e crei il tuo Caddyfile

```
mkdir -p /miahome/caddy/data
nano /miahome/caddy/Caddyfile
```

Poi rimuovi eventualmente il vecchio container e poi crei e lanci quello nuovo

```
docker stop <vecchio_container_caddy>
docker container ls -a
docker rm <container_id_vecchio_caddy>
cd /tmp && curl -LO https://github.com/snakuzzo/caddy-cloudflare-duckdns/archive/refs/heads/latest.zip
unzip latest.zip
cd caddy-cloudflare-duckdns-latest/     
docker build . -t caddy-duckdns:latest
docker run --init -d --name="caddy-duckdns" -e "TZ=Europe/Rome" -v /miahome/caddy/Caddyfile:/etc/caddy/Caddyfile -v /miahome/caddy/data:/data --restart always -p 9123:9123 caddy-duckdns:latest
```

**Reply by snakuzzo** [reply]
Per la cronaca...ho appena rifatto un test in locale e funziona tutto

**Reply by skybarberom** [reply]
Sorry per l'altro thread, non mi sono accorto di averlo scritto nel box sbagliato! Ho rifatto un po' tutto, forse non dandogli un nome alla build non prendevo quella corretta. Aggiungendo il nome come hai indicato
```
docker build . -t caddy-duckdns:latest
```
è comparsa una terza immagine con quel nome ed ho usato quella. Tutto ha funzionato a parte un typo nel caddyfile che ho corretto.
Il container l'ho creato e lanciato però da interfaccia grafica e non da linea di comando. 

Grazie per il supporto!

**Comment by skybarberom** [comment]


**Reply by snakuzzo** [reply]
Però deciditi, cavolo! :)
O usi l'immagine direttamnete o fai la build...non è che vai a caso.
Ti ho risposto su...e ti chiederei la cortesia di non aprire più thread (come stai facendo), ma di fare reply SOTTO L'ULTIMO MESSAGGIO!!!!
